module/VuFind/src/VuFind/Http/CachingDownloader.php

```diff
@@ -0,0 +1,179 @@
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+class CachingDownloader
+{
+    /**
+     * HTTP client
+     *
+     * @var Client
+     */
+    protected $client;
+
+    /**
+     * CacheManager to update caches if necessary.
+     *
+     * @var CacheManager
+     */
+    protected $cacheManager;
+
+    /**
+     * Stored client options for cache key generation.
+     *
+     * @var array
+     */
+    protected $cacheOptions = [];
+
+    /**
+     * Cache to use for downloads
+     *
+     * @var StorageInterface
+     */
+    protected $cache;
+
+    /**
+     * These result modes indicate whether the responses should be decoded.
+     */
+    public const RESULT_MODE_RAW = 'RAW';
+    public const RESULT_MODE_JSON = 'JSON';
+
+    /**
+     * Constructor
+     *
+     * @param Client $client       HTTP client
+     * @param string $cacheManager Base directory for cache
+     */
+    public function __construct(Client $client, CacheManager $cacheManager)
+    {
+        $this->client = $client;
+        $this->cacheManager = $cacheManager;
+        $this->setCacheId('downloader');
+    }
+
+    /**
+     * Set a different cache.
+     *
+     * @param string $cacheId Cache ID
+     *
+     * @return void
+     */
+    public function setCacheId($cacheId)
+    {
+        $cacheName = $this->cacheManager->addDownloaderCache($cacheId);
+        $this->cache = $this->cacheManager->getCache($cacheName);
+    }
+
+    /**
+     * Set client options (HTTP headers and so on).
+     *
+     * @param array|Traversable $options Client options (e.g. HTTP headers)
+     *
+     * @return void
+     */
+    public function setClientOptions($options)
+    {
+        $this->client->setOptions($options);
+
+        if ($options instanceof Traversable) {
+            $options = ArrayUtils::iteratorToArray($options);
+        }
+        $this->cacheOptions = array_merge($this->cacheOptions, $options);
+        ksort($this->cacheOptions);
+    }
+
+    /**
+     * Download a resorce using the cache in the background.
+     *
+     * @param string $url        URL
+     * @param string $resultMode See self::RESULT_MODE_... constants
+     *
+     * @return string
+     */
+    public function download($url, $resultMode = self::RESULT_MODE_RAW)
+    {
+        $cacheItemKey = md5($url);
+        foreach ($this->cacheOptions as $optionKey => $optionValue) {
+            $cacheItemKey .= '#' . $optionKey . '§' . $optionValue;
+        }
+
+        // Return item if exists in cache
+        if ($this->cache->hasItem($cacheItemKey)) {
+            $body = $this->cache->getItem($cacheItemKey);
+            return $this->decodeBody($body, $resultMode, $url);
+        }
+
+        // Add new item to cache if not exists
+        $response = $this->client->setUri($url)->send();
+        if (!$response->isOk()) {
+            throw new \Exception('Could not resolve URL: ' . $url);
```

Comment:
This message could be misleading. Also, should probably catch exceptions as well.